### ISSUE TITLE
Keyset pagination with limit + before is not compatible with GraphQL Relay specification

### ISSUE THREAD

**Issue by dolfinus** [ash-project/ash#333]
**Describe the bug**

Keyset pagination in Ash is not consistent with Relay specification. Specification [requires](https://relay.dev/graphql/connections.htm#sec-Edge-order) `When before: cursor is used, the edge closest to cursor must come last in the result edges`, but Ash returns records from the beginning of the page instead of end.

The specification also says: `The ordering of edges should be the same when using first/after as when using last/before, all other arguments being equal. It should not be reversed when using last/before`.

**To Reproduce**
```elixir
defmodule AshGraphql.Test.RelayTag do
  use Ash.Resource,
    data_layer: Ash.DataLayer.Ets

  actions do
    defaults([:create, :update, :destroy, :read])

    read :read_paginated do
      pagination(required?: true, keyset?: true, countable: true)
    end
  end

  attributes do
    uuid_primary_key(:id)

    attribute(:name, :string)
  end

  identities do
    identity(:name, [:name], pre_check_with: AshGraphql.Test.Api)
  end
end

defmodule AshGraphql.Test.Registry do
  @moduledoc false
  use Ash.Registry

  entries do
    entry(AshGraphql.Test.RelayTag)
  end
end


defmodule AshGraphql.Test.Api do
  @moduledoc false

  use Ash.Api

  # graphql do
  #  authorize? true
  #  debug? false
  #  show_raised_errors? true
  # end

  resources do
    registry(AshGraphql.Test.Registry)
  end
end



letters = ["a", "b", "c", "d", "e", "f", "g"]

for name <- letters do
  tag =
    AshGraphql.Test.RelayTag
    |> Ash.Changeset.for_create(
      :create,
      [name: name]
    )
    |> AshGraphql.Test.Api.create!()
end

{:ok, page1} = AshGraphql.Test.Api.read(%Ash.Query{resource: AshGraphql.Test.RelayTag, sort: [name: :asc], select: [:name]}, action: :read_paginated, page: [limit: 4])

# let's get Keyset page using limit + after
last_record_page1 = page1.results |> Enum.at(3)

# %AshGraphql.Test.RelayTag{
#   __meta__: #Ecto.Schema.Metadata<:built, "">,
#   __metadata__: %{keyset: "g2wAAAABbQAAAAFkag%3D%3D", selected: [:name]},
#   __order__: nil,
#   aggregates: %{},
#   calculations: %{},
#   id: "af77c2d0-f626-4860-af1c-6c629b9e4f3e",
#   name: "d"
# }

{:ok, page2} = AshGraphql.Test.Api.read(%Ash.Query{resource: AshGraphql.Test.RelayTag, sort: [name: :asc], select: [:name]}, action: :read_paginated, page: [limit: 2, before: last_record_page1.__metadata__.keyset])

# returned first 2 records matching the expression `id < before.id`
# not last 2 records

#{:ok,
# %Ash.Page.Keyset{
#   after: nil,
#   before: "g2wAAAABbQAAAAFkag%3D%3D",
#   count: nil,
#   limit: 2,
#   more?: true,
#   rerun: {#Ash.Query<
#      resource: AshGraphql.Test.RelayTag,
#      sort: [name: :asc],
#      limit: 3,
#      select: [:name]
#    >,
#    [
#      page: [limit: 2, before: "g2wAAAABbQAAAAFkag%3D%3D"],
#      verbose?: false,
#      actor: nil,
#      authorize?: false,
#      return_query?: false
#    ]},
#   results: [
#     %AshGraphql.Test.RelayTag{
#       __meta__: #Ecto.Schema.Metadata<:built, "">,
#       __metadata__: %{keyset: "g2wAAAABbQAAAAFhag%3D%3D", selected: [:name]},
#       __order__: nil,
#       aggregates: %{},
#       calculations: %{},
#       id: "a7f27d1f-7afd-46c5-92ef-fc2294224c6a",
#       name: "a"
#     },
#     %AshGraphql.Test.RelayTag{
#       __meta__: #Ecto.Schema.Metadata<:built, "">,
#       __metadata__: %{keyset: "g2wAAAABbQAAAAFiag%3D%3D", selected: [:name]},
#       __order__: nil,
#       aggregates: %{},
#       calculations: %{},
#       id: "26716255-125c-4dc3-86f1-9d6a50fb4730",
#       name: "b"
#     }
#   ]
# }}
```

**Expected behavior**
Last query from example above should return records "b" and "c" instead of "a" and "b", because they are the nearest records before "d".

**Runtime**
 - Elixir 1.13.2 (compiled with Erlang/OTP 24)
 - Erlang/OTP 24 [erts-12.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]
 - Linux 5.18.0-1-MANJARO
 - 1.52.0-rc.9

**Additional context**
Found during implementation of https://github.com/ash-project/ash_graphql/pull/36.

It looks like `after` and `before` share the same implementation:
* `after` filters records using expression `id > cursor.id` with specified sorting
* `before` filters records using expression `id < cursor.id`, with the same sorting
https://github.com/ash-project/ash/blob/ebf25bd6976038333baf0ae38b978c5b713dc7aa/lib/ash/page/keyset.ex#L119-L130

But `before` should instead fetch items from the datalayer with specified limit and reverse sorting order, and then sort results back to specified order.

**Comment by zachdaniel** [comment]
🔥🔥🔥this is perfect. There is a lot in there to address, but the combination of all of those keyset issues ought to provide us everything we need to get this out the door. I'll go through them and add feedback and/or resolve them starting tomorrow. Excellent work!

**Comment by dolfinus** [comment]
Actually, maybe Ash should not initially return pages compatible with GraqhQL Relay (it could break backward compatibility), but at least provide the way to get the last page for keyset pagination.

Like:
```elixir
{:ok, page2} = AshGraphql.Test.Api.read(%Ash.Query{resource: AshGraphql.Test.RelayTag, sort: [name: :asc], select: [:name]}, action: :read_paginated, page: [limit: 2, before: last_record_page1.__metadata__.keyset])


{:ok, last_page} = AshGraphql.Test.Api.page(page2, :last)
```

Offset pagination already support this:
https://github.com/ash-project/ash/blob/9b5c713bcda7de37b74ea47e68c71a03d1e275a7/lib/ash/api/api.ex#L764-L769

**Comment by zachdaniel** [comment]
So I'm pushing a fix that properly orders keyset paginated results when using `before`. Looking(briefly) at the relay specification, it seems like what we have should be sufficient to support their cursors now, right? i.e `last: 10, before: "1"` equals `before: "1", limit: 10`. I can't tell from the spec at this point wether or not we need to support `last` *without* `before`, but if we do we can fake that on the graphql side by reversing the sort order and then reversing the results potentially. Right now, a keyset is only generated on a resource when it is requested for pagination, primarily because the keyset is based on the sort order whereas after #255 is done, the keyset will be based on just the primary key.

**Comment by zachdaniel** [comment]
Alright, this sohuld be fixed in: 6d0cbec24498640e8e6337059371d97dcf84b819
